File: remote/src/main/java/com/kelsos/mbrc/ui/fragments/browse/BrowseAlbumFragment.java
package com.kelsos.mbrc.ui.fragments.browse;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.adapters.AlbumAdapter;
import com.kelsos.mbrc.adapters.EndlessGridRecyclerViewScrollListener;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.Album;
import com.kelsos.mbrc.presenters.BrowseAlbumPresenter;
import com.kelsos.mbrc.ui.fragments.profile.AlbumTracksActivity;
import com.kelsos.mbrc.ui.views.BrowseAlbumView;
import java.util.List;
import roboguice.RoboGuice;

public class BrowseAlbumFragment extends Fragment
    implements AlbumAdapter.MenuItemSelectedListener, BrowseAlbumView {

  @Bind(R.id.album_recycler) RecyclerView recyclerView;
  @Inject private AlbumAdapter adapter;
  @Inject private BrowseAlbumPresenter presenter;
  private EndlessGridRecyclerViewScrollListener scrollListener;

  @NonNull public static BrowseAlbumFragment newInstance() {
    return new BrowseAlbumFragment();
  }

  @Nullable @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    final View view = inflater.inflate(R.layout.ui_library_grid, container, false);
    ButterKnife.bind(this, view);
    RoboGuice.getInjector(getContext()).injectMembers(this);
    presenter.bind(this);
    GridLayoutManager layoutManager = new GridLayoutManager(getContext(), 2);
    scrollListener = new EndlessGridRecyclerViewScrollListener(layoutManager) {
      @Override public void onLoadMore(int page, int totalItemsCount) {
        presenter.load(page);
      }
    };
    recyclerView.setLayoutManager(layoutManager);
    recyclerView.setAdapter(adapter);
    adapter.setMenuItemSelectedListener(this);
    presenter.load();
    return view;
  }

  @Override public void onResume() {
    super.onResume();
    recyclerView.addOnScrollListener(scrollListener);
  }

  @Override public void onPause() {
    super.onPause();
    recyclerView.removeOnScrollListener(scrollListener);
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
  }

  @Override public void onMenuItemSelected(MenuItem item, Album album) {
    switch (item.getItemId()) {
      case R.id.popup_album_tracks:
        openProfile(album);
        break;
      case R.id.popup_album_play:
        presenter.queue(album, Queue.NOW);
        break;
      case R.id.popup_album_queue_last:
        presenter.queue(album, Queue.LAST);
        break;
      case R.id.popup_album_queue_next:
        presenter.queue(album, Queue.NEXT);
        break;
      case R.id.popup_album_playlist:
        break;
      default:
        break;
    }
  }

  @Override public void onItemClicked(Album album) {
    openProfile(album);
  }

  private void openProfile(Album album) {
    Intent intent = new Intent(getContext(), AlbumTracksActivity.class);
    Bundle bundle = new Bundle();
    bundle.putLong(AlbumTracksActivity.ALBUM_ID, album.getId());
    intent.putExtras(bundle);
    startActivity(intent);
  }

  @Override public void updateData(List<Album> data) {
    adapter.updateData(data);
  }

  @Override public void clearData() {
    adapter.clearData();
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/fragments/browse/BrowseAlbumFragment.kt
package com.kelsos.mbrc.ui.fragments.browse

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import butterknife.BindView
import butterknife.ButterKnife
import com.kelsos.mbrc.R
import com.kelsos.mbrc.adapters.AlbumEntryAdapter
import com.kelsos.mbrc.adapters.EndlessGridRecyclerViewScrollListener
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.domain.Album
import com.kelsos.mbrc.presenters.BrowseAlbumPresenter
import com.kelsos.mbrc.ui.fragments.profile.AlbumTracksActivity
import com.kelsos.mbrc.ui.views.BrowseAlbumView
import toothpick.Toothpick
import javax.inject.Inject

class BrowseAlbumFragment : Fragment(), AlbumEntryAdapter.MenuItemSelectedListener, BrowseAlbumView {

  @BindView(R.id.album_recycler) internal lateinit var recyclerView: RecyclerView
  @Inject lateinit var adapter: AlbumEntryAdapter
  @Inject lateinit var presenter: BrowseAlbumPresenter
  private var scrollListener: EndlessGridRecyclerViewScrollListener? = null

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    val view = inflater!!.inflate(R.layout.ui_library_grid, container, false)
    ButterKnife.bind(this, view)
    val scope = Toothpick.openScopes(context.applicationContext, this)
    Toothpick.inject(this, scope)

    presenter.attach(this)
    val layoutManager = GridLayoutManager(context, 2)
    scrollListener = object : EndlessGridRecyclerViewScrollListener(layoutManager) {
      override fun onLoadMore(page: Int, totalItemsCount: Int) {
        presenter.load(page)
      }
    }
    recyclerView.layoutManager = layoutManager
    recyclerView.adapter = adapter
    adapter.setMenuItemSelectedListener(this)
    presenter.load()
    return view
  }

  override fun onDestroy() {
    super.onDestroy()
    Toothpick.closeScope(this)
  }


  override fun onResume() {
    super.onResume()
    recyclerView.addOnScrollListener(scrollListener)
  }

  override fun onPause() {
    super.onPause()
    recyclerView.removeOnScrollListener(scrollListener)
    presenter.detach()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
  }

  override fun onMenuItemSelected(menuItem: MenuItem, album: Album): Boolean {
    when (menuItem.itemId) {
      R.id.popup_album_tracks -> openProfile(album)
      R.id.popup_album_play -> presenter.queue(album, Queue.NOW)
      R.id.popup_album_queue_last -> presenter.queue(album, Queue.LAST)
      R.id.popup_album_queue_next -> presenter.queue(album, Queue.NEXT)
      R.id.popup_album_playlist -> {
      }
      else -> {
        return false
      }
    }

    return true
  }

  override fun onItemClicked(album: Album) {
    openProfile(album)
  }

  private fun openProfile(album: Album) {
    val intent = Intent(context, AlbumTracksActivity::class.java)
    val bundle = Bundle()
    bundle.putLong(AlbumTracksActivity.ALBUM_ID, album.id)
    intent.putExtras(bundle)
    startActivity(intent)
  }

  override fun updateData(data: List<Album>) {
    adapter.updateData(data)
  }

  override fun clearData() {
    adapter.clearData()
  }

  companion object {

    fun newInstance(): BrowseAlbumFragment {
      return BrowseAlbumFragment()
    }
  }
}
